Refactoring of value types

sungam3r opened the pull request:
Final preparations to changes in the main repo.

Goal: get rid of `CoreToVanillaConverter` and all bits from `GraphQL.Language.AST` namespace
Idea: we don't need converting parser AST into GraphQL.NET AST since parser API now has all necessary tools.

UPDATES:
1. Added: IHasValueNode interface
2. Added: ctors with CLR objects AND with ROM for "primitive" value nodes + for comment and description
3. Removed: GraphQLListValue.Value
4. Added: GraphQLName operators

sungam3r commented on src/GraphQLParser.Tests/LexerTests.cs:
```diff
@@ -1055,174 +453,4 @@ public void Lex_BlockString_Tests(int number, string input, string expected)
-        return "\"str\"".Lex();
-    }
-
-    private static Token GetSingleDecimalIntTokenLexer()
```
I merge each group of 4 tests (kind, start, end, value) into one test.

sungam3r commented on src/GraphQLParser.Tests/LexerTests.cs:
```diff
@@ -362,657 +159,258 @@ public void Lex_VeryLongStringToken(bool escape)
 
     [Fact]
-    public void Lex_RightBracketPunctuation_HasCorrectStart()
+    public void NameWithHyphen()
```
One new test.

sungam3r commented on src/GraphQLParser/AST/GraphQLName.cs:
```diff
@@ -47,7 +46,7 @@ public ROM Value
     /// <br/>
     /// This property allocates the string on the heap on first access
-    /// and then caches it until <see cref="Value"/> does not change.
+    /// and then caches it as long as <see cref="Value"/> does not change.
```
fixed

Shane32 commented:
There’s a number of changes in the last few pull requests that cache clr values and strings. We need to take into account the increased memory use here versus before these cached values were added when considering any benefits attained in the main repo.

sungam3r commented on src/GraphQLParser/AST/Values/GraphQLValue.cs:
```diff
@@ -7,4 +7,8 @@ namespace GraphQLParser.AST;
+    /// <summary>
+    /// Returns the CLR object that represents the value of this AST node.
+    /// </summary>
+    public abstract object? ClrValue { get; }
```
All value nodes now can return CLR representation of the parsed value. Some nodes have backing fields, some not (NullValue, BooleanValue, EnumValue).

sungam3r replied:
+NameNode of course (not a value node)

sungam3r replied:
Reverted.

sungam3r commented on src/GraphQLParser/Shim/Numbers.cs:
```diff
@@ -0,0 +1,80 @@
+/// <summary>
+/// Adapter to unify usages of int.TryParse(ReadOnlySpan) for netstandard2.0 and netstandard2.1
+/// </summary>
+internal static class Int
```
Copied without changes from the main repo.

sungam3r replied:
Reverted.

sungam3r commented on src/GraphQLParser/Shim/DecimalData.cs:
```diff
@@ -0,0 +1,29 @@
+// by to produce the Decimal value; bits 24-30 are unused and must be zero;
+// and finally bit 31 indicates the sign of the Decimal value, 0 meaning
+// positive and 1 meaning negative.
+internal readonly struct DecimalData
```
Copied without changes from the main repo.

sungam3r replied:
reverted

sungam3r commented on src/GraphQLParser/ParserContext.Parse.cs:
```diff
@@ -999,6 +999,7 @@ private GraphQLValue ParseObjectValue(bool isConstant)
 
         val.Comment = GetComment();
         val.Fields = ZeroOrMore(TokenKind.BRACE_L, isConstant ? constant : value, TokenKind.BRACE_R);
+        val.Value = _source.Slice(start, _currentToken.End - start - 1);
```
Small fix here - `GraphQLListValue` had `Value` property, but `GraphQLObjectValue` did not.

sungam3r replied:
reverted

sungam3r commented on src/GraphQLParser/AST/GraphQLIntValue.cs:
```diff
@@ -8,13 +11,72 @@ namespace GraphQLParser.AST;
+    /// This property allocates the string on the heap on first access
+    /// and then caches it as long as <see cref="Value"/> does not change.
+    /// </summary>
+    private /*public*/ object TypedValue //TODO: ??? no typed value :(
```
@Shane32 Please pay attention to `IntValue` and `FloatValue`. I did so, because the other options were even worse.

sungam3r replied:
`Parse` methods were ported from the main repo.

sungam3r replied:
Reverted

sungam3r commented:
I understand. These cached values are not always used. Winning is achieved through the complete elimination of the conversion of one AST to another. The number of objects created during conversion is more than the number of strings created for AST **value** nodes (note that only value nodes).

sungam3r commented:
I specifically marked all the places in the main repo that allocate strings and work on the elimination of these allocations. While everything goes according to plan. Somewhere without any problems, it is possible to avoid allocations completely.

Shane32 commented on src/GraphQLParser/AST/GraphQLIntValue.cs:
```diff
@@ -8,13 +11,72 @@ namespace GraphQLParser.AST;
+    }
+
+    /// <inheritdoc />
+    public override object? ClrValue => TypedValue;
```
```suggestion
    public override object? ClrValue => _number ??= TypedValue;
```

Shane32 replied:
This should inline

sungam3r replied:
👍

Shane32 commented on src/GraphQLParser/AST/GraphQLIntValue.cs:
```diff
@@ -8,13 +11,72 @@ namespace GraphQLParser.AST;
+    {
+        get
+        {
+            return _value.Length == 0
+                ? throw new InvalidOperationException("Invalid number (empty string)")
+                : _number ??= Parse();
+
+            object Parse()
+            {
```
```suggestion
            if (_value.Length == 0)
                throw new InvalidOperationException("Invalid number (empty string)");

```

Shane32 commented on src/GraphQLParser/AST/GraphQLIntValue.cs:
```diff
@@ -8,13 +11,72 @@ namespace GraphQLParser.AST;
+
+                // Since BigInteger can contain any valid integer (arbitrarily large), this is impossible to trigger via an invalid query
+                throw new InvalidOperationException($"Invalid number {Value}");
+            }
```
```suggestion
```
Needs reformatting also of course.

Shane32 commented on src/GraphQLParser/AST/GraphQLFloatValue.cs:
```diff
@@ -8,13 +10,74 @@ namespace GraphQLParser.AST;
+    {
+        get
+        {
+            return _value.Length == 0
+                ? throw new InvalidOperationException("Invalid number (empty string)")
+                : _number ??= Parse();
+
+            object Parse()
+            {
```
```suggestion
            if (_value.Length == 0)
                throw new InvalidOperationException("Invalid number (empty string)");

```

Shane32 commented on src/GraphQLParser/AST/GraphQLFloatValue.cs:
```diff
@@ -8,13 +10,74 @@ namespace GraphQLParser.AST;
+                        return dec;
+                }
+
+                return dbl;
+            }
```
```suggestion
            return dbl;
```

Shane32 replied:
Needs reformatting

Shane32 commented on src/GraphQLParser/AST/GraphQLFloatValue.cs:
```diff
@@ -8,13 +10,74 @@ namespace GraphQLParser.AST;
+    }
+
+    /// <inheritdoc />
+    public override object? ClrValue => TypedValue;
```
```suggestion
    public override object? ClrValue => _number ??= TypedValue;
```
Should inline

Shane32 commented on src/GraphQLParser/AST/GraphQLObjectValue.cs:
```diff
@@ -7,10 +8,58 @@ namespace GraphQLParser.AST;
+    }
+
+    /// <inheritdoc />
+    public override object? ClrValue => TypedValue;
```
```suggestion
    public override object? ClrValue => _object ??= TypedValue;
```
(needs other changes here)

And similar changes throughout, so it inlines.

sungam3r replied:
done

Shane32 reviewed (changes requested):
I would change it so `ClrValue` returns `_value ??= TypedValue` throughout, so it can be inlined.

Shane32 commented:
In general, this design is starting to be at odds with itself as a standalone parser.  I'm not complaining, just pointing out that now we are mixing the code that interprets the values into the parser.  When do we stop?  Keep in mind that each ROM hold a pointer to the exact same originally-parsed array and consumes memory for the array pointer, offset and length.  For example, if every single string is eventually converted from a `ROM` to a `string`, and if we are holding a cached copy of the string, then holding the ROM is pointless.  We can just return a ROM created from the string if the ROM was ever requested, and there is no heap allocation for the conversion.  Likewise, it would save memory for a boolean value to only ever hold a boolean, and requesting a ROM from it would return a static ROM just like we do for boxed booleans.

Bottom line, if we don't use the ROMs outside of this project, then holding the ROMs become a waste of memory, and it is better off to just deserialize the data right away.

Yes there are a few instances where it could be faster to interpret it dynamically based on need.  These are going to be limited to value literals (int, float, string), comments, and that's about it.  Query validation examines each node so even if there was a request with unused operations, it is likely that every node is going to be parsed in some fashion.

sungam3r commented:
It is important to see a complete picture. These changes **can** add some memory, but it is important that they allow you to get rid of **much** more memory losses in the main project. I'm working on it.

sungam3r commented:
I advise you to look at `CoreToVanillaConverter` and think why it is necessary. What does this class do? Why?

sungam3r commented:
> In general, this design is starting to be at odds with itself as a standalone parser. 

I know. This is a compromise to achive the best end-to-end results.

sungam3r commented on src/GraphQLParser/AST/Values/GraphQLValue.cs:
```diff
@@ -0,0 +1,21 @@
+    /// <summary>
+    /// Resets cached <see cref="ClrValue"/>.
+    /// </summary>
+    public abstract void Reset();
```
@Shane32 I added explicit `Reset`. It's simpler. In practice, it is unlikely that it will be necessary to someone, but if the value is cached, I was obliged to provide a reset method.

sungam3r replied:
reverted

sungam3r commented on src/GraphQLParser/ParserContext.Parse.cs:
```diff
@@ -855,7 +855,6 @@ private GraphQLValue ParseListValue(bool isConstant)
 
         val.Comment = GetComment();
         val.Values = ZeroOrMore(TokenKind.BRACKET_L, isConstant ? constant : value, TokenKind.BRACKET_R);
-        val.Value = _source.Slice(start, _currentToken.End - start - 1);
```
`Value` property for both `ListValue` and `ObjectValue` was a mistake. These are composite objects that do not have their own data like any other "top-level" AST nodes. Only "primitive" AST nodes should have `ROM Value { get; set; }` property.

sungam3r replied:
I added `IHasValueNode` interface for all "primitive" nodes.

sungam3r commented:
@Shane32 100% diff coverage, I think I finished here and Pr is ready for review. There are a lot of boring work in the main repository changing scalar parsing.

Shane32 commented:
Merge if you like. Or I can review later tonight. But I think I understand generally what this PR is.

sungam3r commented on src/GraphQLParser/AST/GraphQLName.cs:
```diff
@@ -29,35 +28,27 @@ public GraphQLName(ROM value)
-    /// and then caches it until <see cref="Value"/> does not change.
+    /// and then caches it. Call <see cref="Reset"/> to reset cache when needed.
     /// </summary>
     public string StringValue
```
@Shane32 Change to `TypedValue`? Some value nodes have `TypedValue` property. `GraphQLName` is not a value node though.

Example of changes:
![изображение](https://user-images.githubusercontent.com/21261007/149639720-46bae288-2ea0-4198-ae39-1ef5a0977385.png)

sungam3r replied:
Not actual anymore.

sungam3r commented:
I'll wait. I'm interested in review. I use local builds to reference from the main project. Also note that GraphQL.NET has much more changes than here. I do not remember when the last time I made such edits if ever.

sungam3r commented:
If you want, I can publish PR in GraphQL.NET for you to understand what I'm doing. ~150 tests failed for now.

Shane32 commented:
> If you want, I can publish PR in GraphQL.NET for you to understand what I'm doing. ~150 tests failed for now.

Good idea.

Shane32 commented:
I see there's a number of classes where you have added a `Reset` method and removed the set method.  I fail to understand why you would add the `Reset` method.  It seems counterintuitive.  You seemed to have a good design before, where setting the rom would reset the cache.  Makes sense; why change it?  Now you have to set ROM and call Reset?  I don't see any benefit here, just an added complexity.

I think perhaps these questions should be answered:
- Why isn't there a constructor accepting a ROM?
- Why is there a blank constructor?
- When in practice is ROM set excluding during initialization?
- Why can you set the value at all?

It seems to me that there should be a constructor for these classes that accept a ROM, and `Reset` should be removed, and the set method restored to as it was prior to this PR.

Why don't we make all the classes read-only?  Perhaps utilize the `record` keyword if it makes sense also; I'm not sure.

sungam3r commented:
> Now you have to set ROM and call Reset?

No. No one ever needs to reset cache. Even I.

> Makes sense; why change it? 

Because complex value nodes have no ROM property but have other properties (fields for object value node) that **may** be changed. I don't want to bother with all that property change tracking. Again - in real life no one needs it. So I removed tracking and added explicit methods. 

> Why isn't there a constructor accepting a ROM?

I can add one everywhere though it's not nessesary. 

> Why is there a blank constructor?

For convenience. To be able to create node first. In NodeHelper for example.

> When in practice is ROM set excluding during initialization?

Now nowhere. I doubt if anywhere else in future but who knows.

> Why can you set the value at all?

I don't understand.

> Why don't we make all the classes read-only?

I would not. It will disallow changing AST by node visitors. It's rather powerful feature.

Shane32 commented:
Ok

Shane32 commented:
Ok, but I don't like the answer.  The reason stated for allowing changing values at all is "changing AST by node visitors."  This means it is designed to be used by user code.  Users shouldn't be expected to have to call Reset.

Personally I would add the change tracking back for all of the easy ones, and if you need Reset for ObjectValueNode, so be it.  I guess it doesn't matter too much at the moment.

sungam3r commented:
> for all of the easy ones

I did not want to selectively make decisions ("easy" value node or not "easy" value node). There are nodes with values. Values are somehow materialized to objects. These objects can be stored, they can be changed. Now the caching/reset is simple and understandable. It is unlikely that someone will need to change them at all.

sungam3r commented:
Of course, this is a compromise. At first I tried to inherit from base value nodes. It did not work.

Shane32 commented:
I'm not happy we are boxing all values. We can fix it, but it pushes the problem on to `IntGraphType` and similar, which would just box it there.  So might as well box them here and cache them as boxed values.

Shane32 commented:
We should add some simple code to `GraphQLIntValue` to use pre-boxed values for -1 to 10, or some such range.  At least 0 to 5.

Shane32 commented:
Question: why do we add all these constructors instead of just having `ClrValue` as a writable property?  We could even remove `Reset`.

```
    public class GraphQLIntValue : GraphQLParser.AST.GraphQLValue
    {
        public GraphQLIntValue() { }
        public GraphQLIntValue(byte value) { }
        public GraphQLIntValue(decimal value) { }
        public GraphQLIntValue(short value) { }
        public GraphQLIntValue(int value) { }
        public GraphQLIntValue(long value) { }
        public GraphQLIntValue(System.Numerics.BigInteger value) { }
        public GraphQLIntValue(sbyte value) { }
        public GraphQLIntValue(ushort value) { }
        public GraphQLIntValue(uint value) { }
        public GraphQLIntValue(ulong value) { }
        public override object? ClrValue { get; }
        public override GraphQLParser.AST.ASTNodeKind Kind { get; }
        public GraphQLParser.ROM Value { get; set; }
        public override void Reset() { }
    }```

vs

```
    public class GraphQLIntValue : GraphQLParser.AST.GraphQLValue
    {
        public GraphQLIntValue() { }
        public override object? ClrValue { get; set; }
        public override GraphQLParser.AST.ASTNodeKind Kind { get; }
        public GraphQLParser.ROM Value { get; set; }
    }
```

Shane32 commented on src/GraphQLParser/AST/Values/GraphQLListValue.cs:
```diff
@@ -9,18 +10,53 @@ namespace GraphQLParser.AST;
+                }
+                else
+                {
+                    var list = new List<object?>(Values.Count);
+                    foreach (var value in Values)
+                        list.Add(value.ClrValue);
+                    _list = list;
```
Count is known; this should be an array to eliminate an allocation, especially as arrays implement `IList`.

sungam3r replied:
👍  Moved in main repo.

Shane32 commented on src/GraphQLParser/AST/Values/GraphQLListValue.cs:
```diff
@@ -9,18 +10,53 @@ namespace GraphQLParser.AST;
+    /// and then caches it. Call <see cref="Reset"/> to reset cache when needed.
     /// </summary>
-    public List<GraphQLValue>? Values { get; set; }
+    public IList<object?> TypedValue
```
Why public here and nowhere else?

sungam3r replied:
reverted

Shane32 commented on src/GraphQLParser/AST/Values/GraphQLObjectValue.cs:
```diff
@@ -0,0 +1,99 @@
+    /// This property allocates the string on the heap on first access
+    /// and then caches it. Call <see cref="Reset"/> to reset cache when needed.
+    /// </summary>
+    public IDictionary<string, object?> TypedValue
```
Also here

sungam3r replied:
reverted

Shane32 commented on src/GraphQLParser.ApiTests/GraphQLParser.approved.txt:
```diff
@@ -469,8 +499,12 @@ namespace GraphQLParser.AST
+        public GraphQLStringValue(string value) { }
+        public override object? ClrValue { get; }
         public override GraphQLParser.AST.ASTNodeKind Kind { get; }
+        public string TypedValue { get; }
```
Even here `TypedValue` is exposed, but not `IntValue`, `LongValue`, etc.

sungam3r replied:
reverted

Shane32 commented:
Rather `ClrValue` and `TypedValue`, if there is a good reason for both, I would suggest this design pattern:

For `GraphQLValue`,

```cs
public object? ClrValue => GetClrValue();
protected abstract object? GetClrValue();
```

For value nodes such as `GraphQLStringValue`:

```cs
protected override object? GetClrValue() => ClrValue;
public new string ClrValue => _value ??= TypedValue;
```

Now `GraphQLValue.ClrValue` will return an `object?` while `GraphQLStringValue.ClrValue` will return a `string?`.

Not necessary but just an idea.

Shane32 commented:
I guess an int scalar could return one of a few different types, such as `int` or `long` or `BigInteger`.  As such, it wouldn't make sense for it to have a `TypedValue` property.  Ok that makes sense.

Shane32 commented:
I might recommend that we limit the amount of types that `GraphQLIntValue` can return.  For example, only `int`, `long`, `ulong` and `BigInteger`.  All other constructors should be removed.  Then it will notably simplify parsing in `IntGraphType`.

sungam3r commented on src/GraphQLParser/AST/Interfaces/IHasValueNode.cs:
```diff
@@ -0,0 +1,21 @@
+    /// <summary>
+    /// Value of AST node represented as <see cref="ROM"/>.
+    /// </summary>
+    public ROM Value { get; set; }
```
😮 ! @Shane32  I accidentally wrote `public` and it works!

sungam3r replied:
https://jeremybytes.blogspot.com/2019/11/c-8-interfaces-public-private-and.html

sungam3r commented on src/GraphQLParser/AST/Values/GraphQLNullValue.cs:
```diff
@@ -6,11 +6,19 @@ namespace GraphQLParser.AST;
+    /// </summary>
+    public GraphQLNullValue()
+    {
+        Value = "null";
```
Default ROM that may be overridden by parser code. Added for convinience.

sungam3r replied:
I removed setter for `ROM Value`.

sungam3r commented on src/GraphQLParser.Tests/Values/GraphQLFloatValueTests.cs:
```diff
@@ -0,0 +1,47 @@
+    {
+        var value = new GraphQLFloatValue(1.1f);
+        value.Value.Length.ShouldBe(16);
+        value.Value.ShouldBe("1.10000002384186");
```
@Shane32 What do you think ?

sungam3r replied:
Thanks.

sungam3r commented on src/GraphQLParser.Tests/Values/GraphQLIntValueTests.cs:
```diff
@@ -0,0 +1,104 @@
+    public void DecimalBad()
+    {
+        var ex = Should.Throw<ArgumentOutOfRangeException>(() => new GraphQLIntValue(12.3m));
+        ex.Message.Replace(",", ".").ShouldStartWith("Invalid integer number 12.3");
```
> Replace(",", ".")

ru-RU culture locally

Shane32 replied:
I suggest we remove the decimal constructor.  The decimal data type is designed for holding floating point data.  If someone wants to create an integer value from a decimal number, it is their job to store it in a valid data type first; int, long, BigInteger, etc.

sungam3r replied:
How to deal with it here?
```c#
    public virtual GraphQLValue? ToAST(object? value)
        {
            var serialized = Serialize(value);
            return serialized switch
            {
                bool b => new GraphQLBooleanValue(b),
                byte b => new GraphQLIntValue(b),
                sbyte sb => new GraphQLIntValue(sb),
                short s => new GraphQLIntValue(s),
                ushort us => new GraphQLIntValue(us),
                int i => new GraphQLIntValue(i),
                uint ui => new GraphQLIntValue(ui),
                long l => new GraphQLIntValue(l),
                ulong ul => new GraphQLIntValue(ul),
                BigInteger bi => new GraphQLIntValue(bi),
                decimal d => Math.Truncate(d) == d ? new GraphQLIntValue(d) : new GraphQLFloatValue(d), <--------
                float f => new GraphQLFloatValue(f),
                double d => new GraphQLFloatValue(d),
                string s => new GraphQLStringValue(s),
                null => new GraphQLNullValue(),
                _ => throw new NotImplementedException($"Please override the '{nameof(ToAST)}' method of the '{GetType().Name}' scalar to support this operation.")
            };
        }
```

Shane32 replied:
Simple:

`decimal d => new GraphQLFloatValue(d),`

There is nothing incorrect about this.

sungam3r commented:
> Merge if you like. Or I can review later tonight. But I think I understand generally what this PR is.

I'm glad I decided to wait. I reverted majority of my changes. Main repo works fine without `ClrValue`. Please review again.

sungam3r commented:
I'lll add a few tests for `GraphQLName` conversion a bit later to achive 100% diff coverage.

Shane32 commented on src/GraphQLParser.Tests/Values/GraphQLFloatValueTests.cs:
```diff
@@ -0,0 +1,47 @@
+        value.Value.Length.ShouldBe(0);
+    }
+
+    [Fact]
+    public void NanValue()
+    {
+        var ex = Should.Throw<ArgumentOutOfRangeException>(() => new GraphQLFloatValue(double.NaN));
+        ex.Message.ShouldStartWith("Value cannot be NaN.");
+    }
```
Duplicate for Infinity and NegativeInfinity

sungam3r replied:
done 30 seconds ago :)

Shane32 reviewed (changes requested):
I completed a full review of this PR

sungam3r commented on src/GraphQLParser/AST/Values/GraphQLIntValue.cs:
```diff
@@ -0,0 +1,150 @@
+        Value = value.ToString(CultureInfo.InvariantCulture);
+    }
+
+    /// <summary>
+    /// Creates a new instance with the specified value.
+    /// </summary>
+    public GraphQLIntValue(decimal value)
+    {
+        if (Math.Truncate(value) != value)
+            throw new ArgumentOutOfRangeException(nameof(value), $"Invalid integer number {value}");
+        Value = value.ToString(CultureInfo.InvariantCulture);
+    }
```
See ` public virtual GraphQLValue? ScalarGraphType.ToAST(object? value)`. What to do here?

Shane32 replied:
It should create a `GraphQLFloatValue`.  A decimal is a floating-point number.

sungam3r replied:
OK.

sungam3r replied:
done

sungam3r commented on src/GraphQLParser/AST/Values/GraphQLBooleanValue.cs:
```diff
@@ -6,11 +6,26 @@ namespace GraphQLParser.AST;
+    /// </summary>
+    public GraphQLBooleanValue(bool value)
+    {
+        Value = value ? "true" : "false";
```
🤔

sungam3r replied:
```c#
public ROM Value
{
    get => "null";
    set => { }
}
```

for NullValue it work since 'null' is constant.

Shane32 replied:
Then leave it alone.

Shane32 commented on src/GraphQLParser/AST/Values/GraphQLFloatValue.cs:
```diff
@@ -0,0 +1,109 @@
+    /// </summary>
+    public ROM Value { get; set; }
+
+    private static double ValidateValue(double value)
+    {
+        // TODO: see https://github.com/graphql-dotnet/graphql-dotnet/pull/2379#issuecomment-800828568 and https://github.com/graphql-dotnet/graphql-dotnet/pull/2379#issuecomment-800906086
+        // if (double.IsNaN(value) || double.IsInfinity(value))
+        if (double.IsNaN(value))
+            throw new ArgumentOutOfRangeException(nameof(value), "Value cannot be NaN."); // Value cannot be NaN or Infinity.
+
+        return value;
+    }
```
See https://spec.graphql.org/draft/#sec-Float

> Non-finite floating-point internal values (NaN and Infinity) cannot be coerced to Float and must raise a field error.

Similar text about input coercion.  It's pretty explicit that NaN and Infinity are not supported by GraphQL.

sungam3r replied:
done

sungam3r commented on src/GraphQLParser/AST/Values/GraphQLIntValue.cs:
```diff
@@ -0,0 +1,150 @@
+    public GraphQLIntValue(byte value)
+    {
+        Value = value.ToString(CultureInfo.InvariantCulture);
+    }
```
done

Shane32 commented on src/GraphQLParser/AST/Values/GraphQLBooleanValue.cs:
```diff
@@ -6,15 +6,25 @@ namespace GraphQLParser.AST;
 /// </summary>
 [DebuggerDisplay("GraphQLBooleanValue: {Value}")]
-public class GraphQLBooleanValue : GraphQLValue
+public class GraphQLBooleanValue : GraphQLValue, IHasValueNode
```
Now that we do not allow changing a boolean, we could have `GraphQLBooleanValue` abstract, and have separate derived classes for true and false variations.  I don't really care either way but it could possibly save some tiny bit of memory.

Perhaps separate PR even if we do want to do that.

sungam3r replied:
I did not do this because there will be combinations with descendant internal classes . In any case, if desired, this can be done later.

sungam3r replied:
#195

Shane32 commented on src/GraphQLParser/AST/Values/GraphQLBooleanValue.cs:
```diff
@@ -6,15 +6,25 @@ namespace GraphQLParser.AST;
     /// Boolean value represented as <see cref="ROM"/>.
     /// </summary>
-    public ROM Value { get; set; }
+    public ROM Value => _value ? "true" : "false";
```
I think we should add a property to return the underlying boolean.  Saves another comparison in the main project.

sungam3r replied:
done

Shane32 reviewed (approved):
I'm okay either way on adding the clr property for BooleanValue.  All files reviewed.  (well, non-test files)
